### ISSUE TITLE
List based calls don't set RequestID

### ISSUE THREAD

**Issue by jbonzo** [stripe/stripe-go#1141]
Go version: 1.14.2
Stripe Go version: v71.28.0

I was having an issue when trying to list charges on a customer with many charges where sometimes the return set was null. I added logic to log the stripe request ID to further investigate but saw that the stripe request ID was never being logged. My logging logic was doing a bounds check like so
```golang
if len(charges) > 0 && charges[0].LastResponse != nil {
    // add transactions[0].LastResponse.RequestID to logging context
}
```
and it was never being met.

I entered the debugger and saw that `LastResponse` was nil on every charge in the slice.
I read the client code in the charge package and I think I identified where the `LastResponse` information is being lost.
https://github.com/stripe/stripe-go/blob/v71.28.0/charge/client.go#L79-L86
```golang
// List returns an iterator that iterates all charges.
func (c Client) List(listParams *stripe.ChargeListParams) *Iter {
	return &Iter{stripe.GetIter(listParams, func(p *stripe.Params, b *form.Values) ([]interface{}, stripe.ListMeta, error) {
		list := &stripe.ChargeList{}
		err := c.B.CallRaw(http.MethodGet, "/v1/charges", c.Key, b, p, list)

		ret := make([]interface{}, len(list.Data))
		for i, v := range list.Data {
			ret[i] = v
		}

		return ret, list.ListMeta, err
	})}
}
```

The value `list` does contain a non nil`LastResponse`, which contains helpful information for debugging, but each item in `list.Data` does not contain such field. As we can see here the only information parsed from `list` are these data items as they are used to construct the `Iter` instance.

I believe this behavior extends to the other list based calls and I'm wondering if this was intentional.

If a change is required I am happy to draft a PR if I could learn more about your contribution process

**Comment by brandur-stripe** [comment]
Thanks @jbonzo!

I thought about solving this by either (1) putting a last response accessible somewhere like `ListMeta`, or (2) just setting it on every value in a list as we iterate through it, but if I'm reading this right, I think that making either of those options work might require some pretty significant refactoring because the only place we have access to the original list object (e.g. `ChargeList`) is in the individual `Query` implementations for each resource type. e.g.:

```
func (c Client) List(listParams *stripe.ChargeListParams) *Iter {
	return &Iter{stripe.GetIter(listParams, func(p *stripe.Params, b *form.Values) ([]interface{}, stripe.ListMeta, error) {
		list := &stripe.ChargeList{}
		err := c.B.CallRaw(http.MethodGet, "/v1/charges", c.Key, b, p, list)

                ...

               // We return a list of objections and a `ListMeta`, but `list` itself is discarded ...
		return ret, list.ListMeta, err
	})}
}
```

So we'd have to change every single one of those which isn't great.

The only other thing I can think of is to try and change `APIResource.SetLastResponse` to try do set responses recursively by looking for other `APIResponse` fields within itself. So for a list object, we'd see the array of `APIResource` implementations under `Data`, and call `SetLastResponse` on everyone of them. There would be some performance impact thanks to the reflection, but this might be our best/easiest possible fix.

**Comment by jbonzo** [comment]
@brandur-stripe I think I understand the different options we have. 

Option 1 (adding the attribute to `ListMeta`) is the least desirable because it involves changing the definition of a struct that may have implications down the road that I'm not too familiar with to see.

Option 2 seems better since I think every Stripe resource struct already is an `APIResource` so it has the request ID defined in it. The refactor does seem like a pain but it does feel like a 1 time developer cost, unless you see future pain when adding new Stripe resources and ensuring that the developer adds the request ID logic there as well.

Option 3 sounds a little more complex in development cost but less prone for error and more scalable to new resources. I do see the performance cost due to the reflection, though I would be interested in seeing how bad that performance hit would be.

I'm curious on your thoughts on these tradeoffs. Personally I think option 2 sounds better and I am willing to front that monotonous work as long as you think this option scales for future resources in development.

**Comment by brandur-stripe** [comment]
@jbonzo Yeah, good points.

I was leaning toward option (3) the last time I wrote you because for option (2), along with being a lot of work, it also would mean adding a lot of boilerplate to each `List` implementation we have in the library, and there are dozens of them.

Looking at this again though, I think it would be okay to replace this line:

``` go
err := c.B.CallRaw(http.MethodGet, "/v1/charges", c.Key, b, p, list)
```

In each of those calls with something that's run specifically for lists, and which does some of the "post" list work like assigning that API response.

I've put it on the board this week for me to investigate this option a bit more, and if it looks good, I'll pull the trigger and run it by the rest of the team here.

Thanks for the offer go help do the option (2) work! That is extremely generous, but I think in this case it'd be a big enough refactor though, that it's probably better if you left it to us :)

**Comment by jbonzo** [comment]
@brandonl-stripe Thanks for the great correspondence. I look forward to seeing the solution whenever you all prioritize it.

**Comment by brandur-stripe** [comment]
Okay, I played around initially with extracting a special "list" function that could be called for each of the `GetIter` invocations, but just thanks to the typing around slices (basically how you can't convert a typed slice like `[]*Customer` to an untyped one like `[]interface{}` without reflection), I found that I couldn't really extract much into it _besides_ setting a last response, so it wasn't all that big of a win.

I'm at the point now where I think that the right answer for this might just be going through each `List` function and adding a line to set the last response. For example:

``` diff
diff --git customer/client.go customer/client.go
index ad9140e..857b787 100644
--- customer/client.go
+++ customer/client.go
@@ -78,6 +78,7 @@ func (c Client) List(listParams *stripe.CustomerListParams) *Iter {
 
 		ret := make([]interface{}, len(list.Data))
 		for i, v := range list.Data {
+			v.SetLastResponse(list.LastResponse)
 			ret[i] = v
 		}
```

It's one more line in each `List` function, but all of these are copied and pasted when new ones are added, so it's not clearly a major loss.

I also thought about putting the last response on `Iter`, but that would also require a bunch of refactoring to return a last response that `Iter` could set. Basically this on every `List`:

``` diff
diff --git customer/client.go customer/client.go
index ad9140e..7d9f20e 100644
--- customer/client.go
+++ customer/client.go
@@ -72,7 +72,7 @@ func List(params *stripe.CustomerListParams) *Iter {
 
 // List returns a list of customers.
 func (c Client) List(listParams *stripe.CustomerListParams) *Iter {
-	return &Iter{stripe.GetIter(listParams, func(p *stripe.Params, b *form.Values) ([]interface{}, stripe.ListMeta, error) {
+	return &Iter{stripe.GetIter(listParams, func(p *stripe.Params, b *form.Values) ([]interface{}, stripe.ListMeta, *stripe.APIResponse, error) {
 		list := &stripe.CustomerList{}
 		err := c.B.CallRaw(http.MethodGet, "/v1/customers", c.Key, b, p, list)
 
@@ -81,7 +81,7 @@ func (c Client) List(listParams *stripe.CustomerListParams) *Iter {
 			ret[i] = v
 		}
 
-		return ret, list.ListMeta, err
+		return ret, list.ListMeta, list.LastResponse, err
 	})}
 }
 ```

It would also force you to access last response through an iterator as opposed to through an API resource like everywhere else, so between those two things, that solution really doesn't seem any better to me.

@remi-stripe Any thoughts on just adding `v.SetLastResponse(list.LastResponse)` to everything? (First code block above.) If not, I'm just going to go through and do that.

**Comment by remi-stripe** [comment]
When you list a Customer you get a list of Subscriptions in the `Subscriptions` property and we don't set `LastResponse` on those. Same when you have a Customer with `DefaultSource` expanded. The idea was that we set `LastResponse` only top-level so having it partially set on some objects as part of a list/pagination _feels_ off for some reason in my head. Especially if the goal is to debug something strange specifically versus something actively needed in a normal integration.

Similarly, we have new "list" endpoints that are a bit custom such as `ListLineItems` on `CheckoutSession` so we should make sure those behave the right way too.

Reading through the thread, it seems that adding it to the `Iter` or `ListMeta` is maybe overkill so I wouldn't be opposed to that change if we believe it's useful. Also happy to do the manual work @brandur-stripe if that helps!

**Comment by brandur-stripe** [comment]
> Reading through the thread, it seems that adding it to the `Iter` or `ListMeta` is maybe overkill so I wouldn't be opposed to that change if we believe it's useful. Also happy to do the manual work @brandur-stripe if that helps!

Thanks Remi!

Okay, I looked at this yet _again_, and now have a slightly different version of the proposal that ends up adding access to the list object to `Iter` so that you could access last response with something like this:

``` go
requestID := it.CustomerList().LastResponse.RequestID
```

Honestly, I don't think it's practically speaking all that much better than the last solution I pitched, but I do like that this is more likely to give us a little more future compatibility in case we need to be able to access the original list object for any other reasons.

I put up a sketch in #1148 to show what this would look like.

**Comment by brandur-stripe** [comment]
#1148 released as 71.44.0.
